refactor(cri): restructure CRI API (improve robustness, clarity and maintainability)

incertum opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

I have added very detailed comments to the code diff re the implemented changes.

At a high level the refactor achieves the following:

- Add robustness wrt to metadata retrieved from the PodSandboxStatusResponse
- Improve technical clarity and maintainability (flow and steps are now easy to follow and understand)
- Remove some redundancies
- Expand unit tests (further improving technical clarity and maintainability)
- This PR does NOT add any extra API calls that weren't done prior to this refactor

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/libs/issues/1589

Why this PR matters:
- Currently, we are not handling or enriching pod sandbox container processes correctly. This means adopters can falsely report that a container image is missing, when in fact it may be legitimate. Once this PR is merged, adopters will have the option to compare container and pod sandbox IDs. If they are equal, they will know that the container images will never exist.
- Additionally, I would rate this code cleanup as highly crucial for slightly optimizing the CRI API lookups. More importantly, it will make them unit-testable and improve transparency and maintainability in the long run.

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1600#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -656,85 +694,38 @@ inline std::string cri_interface<api>::get_container_image_id(const std::string
 }
 
-template<typename api>
-inline bool cri_interface<api>::parse_containerd(const typename api::ContainerStatusResponse &status,
```
Removing `parse_containerd` brings significant benefits by enhancing the clarity and transparency of the codebase. This is because the parsing previously done here always also applied for `cri-o` and therefore at best this approach was confusing overall.

Now the main `parse` function calls all parsers and we get rid of this nested additional jump.

Also I have the suspicion we were inefficient with the IP enrichment / lookup before (it may have been called twice?), now the main`parse` function has a very clear structure ensuring we make only one call for container and pod sandbox status responses and parse them right away.

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -656,85 +694,38 @@ inline std::string cri_interface<api>::get_container_image_id(const std::string
+			const auto &resp_pod_sandbox_container = pod_sandbox_status_resp.status();
+			const auto &resp_pod_sandbox_container_info = pod_sandbox_status_resp.info();
+			const auto root_pod_sandbox = get_info_jvalue(resp_pod_sandbox_container_info);
+			parse_cri_base(resp_pod_sandbox_container, container);
```
We previously didn't do a good job in handling pod sandbox containers and the labels (container or pod sandbox labels) were confusing in any aspect. Now each container regardless of whether it is a normal container or a pod sandbox container has `m_labels` and `m_pod_sandbox_labels`. 

Further improvements include adding all info that makes sense. Now pod sandbox container processes are also enriched the best way possible. Note that in the k8s filterchecks we do not need pod sandbox containers anymore to retrieve `m_pod_sandbox_labels`.

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -746,33 +737,27 @@ inline bool cri_interface<api>::parse(const libsinsp::cgroup_limits::cgroup_limi
+	const auto &resp_container = container_status_resp.status();
+	const auto &resp_container_info = container_status_resp.info();
+	const auto root = get_info_jvalue(resp_container_info);
+	parse_cri_base(resp_container, container);
```
No more ad-hoc code here, the main `parse` function consistently calls `parse_*` helpers.

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -782,23 +767,32 @@ inline bool cri_interface<api>::parse(const libsinsp::cgroup_limits::cgroup_limi
 			container.m_id.c_str(), container.m_imagerepo.c_str(), container.m_imagetag.c_str(),
 			container.m_image.c_str(), container.m_imagedigest.c_str());
 
+	// Enabled by default for Falco consumer
```
todo confirm

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -782,23 +767,32 @@ inline bool cri_interface<api>::parse(const libsinsp::cgroup_limits::cgroup_limi
+		const auto &resp_pod_sandbox_container_info = pod_sandbox_status_resp.info();
+		const auto root_pod_sandbox = get_info_jvalue(resp_pod_sandbox_container_info);
+		// Add pod response network and labels to original container
+		parse_cri_pod_sandbox_network(resp_pod_sandbox_container, root_pod_sandbox, container);
```
I mentioned before that we may have done the network enrichment twice before and now only one time in a transparent and clean way.

incertum replied:
@gnosek highlighting again that I am not introducing extra CRI API calls. They were already there before. 

In summary, the design is that for each container that is not a pod sandbox container you get a `ContainerStatsResponse` once and also a `PodSandboxStatusResponse` once and fully make use of the fact that you have the `PodSandboxStatusResponse`.

If we want to complicate things and cache `PodSandboxStatusResponse` info in pod sandbox containers so that the call is only done once for each pod is a different discussion. I would not recommend complicating things for now, but we could think about it in the future.

incertum commented on userspace/libsinsp/sinsp_filtercheck_k8s.cpp:
```diff
@@ -225,8 +225,9 @@ uint8_t* sinsp_filter_check_k8s::extract(sinsp_evt *evt, OUT uint32_t* len, bool
-	if(container_info == nullptr || container_info->m_labels.empty())
+	// No m_pod_sandbox_id means no k8s.
+	// m_pod_sandbox_id retrieved from the ContainerStatusResponse CRI API call.
+	if(container_info == nullptr || container_info->m_pod_sandbox_id.empty())
```
@leogr @Andreagit97 this should satisfy the desired behavior, please confirm.

incertum commented on userspace/libsinsp/sinsp_filtercheck_k8s.cpp:
```diff
@@ -259,69 +263,47 @@ uint8_t* sinsp_filter_check_k8s::extract(sinsp_evt *evt, OUT uint32_t* len, bool
+				m_tstr = container_info->m_pod_sandbox_labels.at(m_argname);
+				RETURN_EXTRACT_STRING(m_tstr);
 			}
-			const sinsp_container_info::ptr_t sandbox_container_info =
```
No more extra pod sandbox container lookups from the container cache (now more efficient). Plus I suspect this may have caused inconsistencies in corner cases as those containers were processed at different times in the async event queue.

incertum commented on userspace/libsinsp/sinsp_filtercheck_k8s.cpp:
```diff
@@ -259,69 +263,47 @@ uint8_t* sinsp_filter_check_k8s::extract(sinsp_evt *evt, OUT uint32_t* len, bool
-			return NULL;
-		}
-		RETURN_EXTRACT_STRING(container_info->m_pod_cniresult);
+		// Requires s_cri_extra_queries enabled, which is the default for Falco.
```
Clear docs.

incertum commented on userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp:
```diff
@@ -53,46 +53,46 @@ std::string container_info_json = R"({
-    "log_path": "busybox.0.log",
-    "linux": {
-        "resources": {
+        "metadata": {
```
only adjusted indentation

incertum commented on userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp:
```diff
@@ -741,4 +690,88 @@ TEST_F(sinsp_with_test_input, container_parser_cri_containerd)
 	ASSERT_EQ(get_field_as_string(evt, "k8s.pod.cni.json"), "{\"bridge\":{\"IPConfigs\":null},\"eth0\":{\"IPConfigs\":[{\"Gateway\":\"10.244.0.1\",\"IP\":\"10.244.0.2\"}]}}");
 }
+
+TEST_F(sinsp_with_test_input, container_parser_cri_containerd_sandbox_container)
```
Now extra tests for pod sandbox containers.

incertum commented on userspace/libsinsp/test/filterchecks/k8s.cpp:
```diff
@@ -194,15 +191,11 @@ TEST_F(sinsp_with_test_input, K8S_FILTER_check_fields_value_with_no_labels)
+	container_info->m_pod_sandbox_cniresult = cni_json;
 	m_inspector.m_container_manager.add_container(std::move(container_info), init_thread_info);
 
-	auto sandbox_container_info = std::make_shared<sinsp_container_info>();
```
This test now also clearly shows that we have all we need in the container, no need for extra pod sandbox container lookup from the container cache anymore.

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -473,26 +499,102 @@ inline bool cri_interface<api>::parse_cri_user_info(const Json::Value &info, sin
+
+	const auto pod_ip = status.network().ip();
+	uint32_t ip;
+	if(pod_ip.empty() || 
```
Let's double check if we can adjust this check for clear behavior ....

incertum commented on userspace/libsinsp/cri.hpp:
```diff
@@ -808,6 +785,27 @@ inline bool cri_interface<api>::parse(const libsinsp::cgroup_limits::cgroup_limi
 		}
+
+		/*
+		* The recent refactor makes full use of PodSandboxStatusResponse, removing the need to access pod sandbox containers
```
@gnosek highlighting again that I am not introducing extra CRI API calls. They were already there before. 

In summary, the design is that for each container that is not a pod sandbox container you get a `ContainerStatsResponse` once and also a `PodSandboxStatusResponse` once and fully make use of the fact that you have the `PodSandboxStatusResponse`.

If we want to complicate things and cache `PodSandboxStatusResponse` info in pod sandbox containers so that the call is only done once for each pod is a different discussion. I would not recommend complicating things for now, but we could think about it in the future.

incertum commented on userspace/libsinsp/sinsp_filtercheck_k8s.cpp:
```diff
@@ -259,69 +263,86 @@ uint8_t* sinsp_filter_check_k8s::extract(sinsp_evt *evt, OUT uint32_t* len, bool
+			if(container_info->m_pod_sandbox_cniresult.empty()) // more robust check than checking for empty labels
 			{
-				sandbox_container_id.resize(12);
+				// Fallback: Retrieve PodSandboxStatusResponse fields stored in pod sandbox container
```
For maximum robustness in retrieving the fields we get from the PodSandboxStatusResponse fetch them either from the container or as fallback from the pod sandbox container.

incertum commented:
@gnosek may I ask for a first-pass review from you? Thanks a bunch in advance?

incertum commented:
If ok with everyone

/milestone 0.15.0

incertum commented:
@falcosecurity/libs-maintainers kindly checking in on the status of this PR. Thank you.

incertum commented:
Please note that this PR has been open for 2 months now without receiving any review whatsoever. During this time, it appears that our focus has been primarily directed towards prioritizing and integrating similarly complex code alterations. 

Why this PR matters:
- Currently, we are not handling or enriching pod sandbox container processes correctly. This means adopters can falsely report that a container image is missing, when in fact it may be legitimate. Once this PR is merged, adopters will have the option to compare container and pod sandbox IDs. If they are equal, they will know that the container images will never exist.
- Additionally, I would rate this code cleanup as highly crucial for slightly optimizing the CRI API lookups. More importantly, it will make them unit-testable and improve transparency and maintainability in the long run.

FedeDP commented:
PR LGTM, but i am in no way an expert on this part of the code :) 
I really like the refactor though, that is the most important thing.
/cc @gnosek :pray:

leogr commented:
On my side, I will prioritize this once https://github.com/falcosecurity/libs/pull/1595 gets merged (hopefully soon).
/assign

I still want to confirm that this is targetted for libs 0.15

incertum commented:
@leogr this PR should be the next one. Rebased and added another commit on top in an attempt to improve comments and naming and code organization even more.

If it help, happy to schedule a call to more interactively review this PR, just let me know, thanks a bunch!

leogr commented:
> @leogr this PR should be the next one. Rebased and added another commit on top in an attempt to improve comments and naming and code organization even more.
> 
> If it help, happy to schedule a call to more interactively review this PR, just let me know, thanks a bunch!

I'm working on it. @therealbobo already did some tests, and everything worked properly (@therealbobo do you confirm?)/

cc @falcosecurity/libs-maintainers for visibility

incertum commented:
> I'm working on it. @therealbobo already did some tests, and everything worked properly (@therealbobo do you confirm?)/

Thank you! While it's a terrible diff, I believe this PR is less risky, plus at least it's backed up my unit tests (acknowledging some limitations w/ mock CRI response unit tests, but still better than before when we had nothing).

Perhaps, double check and pay close attention to all conditional checks etc. 

Plus plz let me know if the filterchecks fallback mechanisms are ok with you:
- We now consistently first try to get all pod sandbox related fields from the container
- Falling back to fetching the separate pod sandbox container from the cache (which is now also fully populated, it was not the case before). The `cniresult` string was chosen as check for it as it should always be there in Kubernetes, while labels could legitimately be empty who knows. perhaps down the road there could be another improvement PR< but I wouldn't aim for absolutely perfect right now either.
- The fallback could be important if for example the pod sandbox CRI call failed after a successful container CRI calls. Probably not very common, but it could happen since we do not mark the overall lookups within `parse` as unsuccessful when the container lookups succeeded but the subsequent additional pod sandbox CRI call failed.

therealbobo commented:
@leogr I did! I also see new commits. Let me check! 👀

incertum commented:
> @leogr I did! I also see new commits. Let me check! 👀

Only the last commit is new, else conflict-free rebase. The last commit is merely one more renaming and adding more comments pass etc, no changes really.

incertum commented:
Since we are doing this, let's do it right: I pushed one more commit to really be super consistent also in the order of calls in the CRI code base (plus few more comments). Ok to call me crazy on those details, but as said hopefully we don't need to touch this again for a bit.

leogr commented:
> Since we are doing this, let's do it right: I pushed one more commit to really be super consistent also in the order of calls in the CRI code base (plus few more comments). Ok to call me crazy on those details, but as said hopefully we don't need to touch this again for a bit.

:+1: Testing this again (in progress).

poiana commented:
LGTM label has been added.  <details>Git tree hash: 0087a2cc73d9fc24d2e741c5bb350dc12e4c7e33</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1600#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1600#pullrequestreview-1910912754" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1600#pullrequestreview-1922229763" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1600#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1600#pullrequestreview-1910912754" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,incertum,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
